### ISSUE TITLE
solve_quartic float16 never takes the cubic fallback: zero_tol=1e-12 underflows to 0

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4498]
`solve_quartic` never takes its cubic fallback for float16 input, because the leading-coefficient tolerance underflows to zero.

`kornia/geometry/solvers/polynomial_solver.py:278` sets `zero_tol = 1e-6 if coeffs.dtype == torch.float32 else 1e-12`, and `:281` tests `torch.abs(a) < zero_tol`. In float16, `1e-12` rounds to `0.0`, so `abs(a) < 0` is never true. A quartic with `a = 0`, which is really a cubic, goes down the Ferrari path and divides by zero.

## Repro

```python
import torch
from kornia.geometry.solvers import solve_quartic

c = [[0.0, 1.0, -6.0, 11.0, -6.0]]          # (x - 1)(x - 2)(x - 3), leading coefficient 0
for dt in (torch.float16, torch.bfloat16, torch.float32):
    print(dt, solve_quartic(torch.tensor(c, dtype=dt))[0].tolist(), torch.tensor(1e-12, dtype=dt).item())
```

## Observed (main `dc0af1a4`, torch 2.14.0, CPU)

```
torch.float16  [nan, nan, nan, nan]               0.0
torch.bfloat16 [3.03125, 1.046875, 1.9375, 0.0]    1.0018652574217413e-12
torch.float32  [3.0, 0.9999998807907104, 2.0, 0.0] 9.999999960041972e-13
```

On a 2000-row family with a near-zero leading coefficient, all 2000 float16 rows fail. The same thing happens on the two open #4346 fixes: #4357 returns `[nan, -inf, nan, nan]` for this literal and #4375 behaves the same. Both promote the Ferrari arithmetic of half-precision inputs to float32, but they still take the fallback decision on the float16 coefficients.

## Expected

The cubic roots `1, 2, 3` in float16, as bfloat16 and float32 already give. The tolerance should be representable in the input dtype, or the fallback decision should be made on the promoted float32 coefficients. #4357 already promotes half inputs, so the second option fits it. It is a separate defect from #4346 (root selection) and #4474 (near-square float32 quartics).

Posted on behalf of @ducha-aiki by Claude (Opus 5).

**Comment by msparsh13** [comment]
[**DOUBT**] While testing the fix for this issue (#4498), I noticed that the existing `test_random `also fails for `float16 `and `bfloat16`, even with the internal quartic computation promoted to `float32`.
- This appears to be separate from the cubic-fallback issue here because the random test uses `a = 1`.
- With the current atol=1e-3, rtol=1e-3, float16 fails with 15/40 mismatched elements in main code itself.
-  After the change,In one run, `float16 `failed with 13/40 mismatched elements and a greatest absolute difference of 0.013.
- In the same run, `bfloat16 `failed with 22/40 mismatched elements and a greatest absolute difference of 0.051.
- Is this level of accuracy expected for half/bfloat16 quartic solving, or should the random test use dtype-specific tolerances?
- I haven't changed the test or tolerances since this seems like a separate numerical-accuracy question.
- I haven't opened a PR for this yet.

**Comment by maheen8q** [comment]
I'm looking into #4498 as well. I understand the float16 fallback issue is separate from the random quartic accuracy failures you've found. Would it be okay if I focused on reproducing the fallback bug and adding a regression test, while leaving the random-test tolerance question separate?

**Comment by ducha-aiki** [comment]
@msparsh13 you were first here, with a fix already in testing, so go ahead and open it. There's no PR yet; since we can't assign outside contributors, this comment is the claim.

On your question: `test_random` at float16/bfloat16 is already a recorded known failure (`testing/half_precision_xfails/cpu_float16.txt` and `cpu_bfloat16.txt`) and is separate from this issue, so leave that test and its tolerances alone. One manifest line does belong to this issue: `TestQuarticSolver::test_solve_quartic[cpu-float16-coeffs3-expected_solutions3]` is exactly the `[0, 1, -6, 11, -6]` literal. Once the fallback works it will XPASS, so delete that line in the same PR. `--verify-known-failures --known-failure-profile=cpu-float16` will show whether any other line flips. #4669 is also open on `polynomial_solver.py`, so a rebase may be needed.

@maheen8q the fallback bug is the part @msparsh13 is fixing. If there's no PR from them by 2026-09-30, it's yours.

Posted on behalf of @ducha-aiki by Claude (Opus 5.5).
